packages/app-lib/src/util/json.rs

```diff
@@ -0,0 +1,42 @@
+use crate::ErrorKind::InputError;
+use crate::Result;
+use serde::de::DeserializeOwned;
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+pub async fn parse_object_async_reader<R, T>(reader: &mut R) -> Result<T>
+where
+    R: AsyncRead + Unpin,
+    T: DeserializeOwned,
+{
+    let first_char = reader.read_u8().await?;
+    if first_char != b'{' {
+        return Err(InputError(format!(
+            "Expected '{{' to start JSON object, but found '{}'",
+            first_char as char
+        ))
+        .into());
+    }
+
+    let mut json_data = vec![first_char];
+    let mut depth = 1usize;
+    while depth > 0 {
+        let char = reader.read_u8().await?;
+        json_data.push(char);
+        match char {
+            b'{' => depth += 1,
+            b'}' => depth -= 1,
+            b'"' => loop {
+                let char = reader.read_u8().await?;
+                json_data.push(char);
+                match char {
+                    b'\\' => json_data.push(reader.read_u8().await?),
+                    b'"' => break,
+                    _ => {}
+                }
+            },
+            _ => {}
+        }
+    }
+
+    Ok(serde_json::from_slice(&json_data)?)
+}
```

Comment:
All of this could be replaced by a single call to `serde_json::from_str` if the messages in the RPC protocol were split by some delimiter... In particular, Tokio's [`LineCodec`](https://docs.rs/tokio-util/latest/tokio_util/codec/index.html#example-decoding-using-linescodec) is a great fit for this.

I know I said on Slack that not using a delimiter was okay, but I wasn't thinking about this unnecessary complexity on the Rust side at the time. The least overall code we have, the better it is to maintain.

In the Java side of things, you could use [`BufferedReader#readLine`](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#readLine--) for the splitting. Writing line-delimited records is trivial enough in both languages, I think.